{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rosalind bioinformatic problem practice\n",
    "#### https://rosalind.info/problems/locations/\n",
    "#### Aine Fairbrother-Browne\n",
    "#### 2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Transcribing DNA into RNA\n",
    "# An RNA string is a string formed from the alphabet containing 'A', 'C', 'G', and 'U'.\n",
    "# Given a DNA string t corresponding to a coding strand, its transcribed RNA string u is \n",
    "# formed by replacing all occurrences of 'T' in t with 'U' in u. \n",
    "\n",
    "def transcribe(dna):\n",
    "\n",
    "    # define dict of nucleotides\n",
    "    transcription_dict = {'A':'A', 'T':'U', 'C':'C', 'G':'G'}\n",
    "\n",
    "    # split input string into list of nucleotides\n",
    "    dna_split = [*dna]\n",
    "\n",
    "    # check whether input dna is less than or equal to 1000 nucleotides\n",
    "    if(len(dna_split)<=1000):\n",
    "\n",
    "        # lookup nucleotides in dict\n",
    "        rna_list = [transcription_dict[x] for x in dna_split]\n",
    "\n",
    "        # join into string and return\n",
    "        rna_string = ''.join(rna_list)\n",
    "        print(rna_string)\n",
    "    else:\n",
    "        print('Sequence longer than the 1000 nucleotide limit.')\n",
    "\n",
    "# test, should give GAUGGAACUUGACUACGUAAAUU\n",
    "transcribe('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Counting DNA nucleotides\n",
    "# A string is simply an ordered collection of symbols selected from some alphabet and formed \n",
    "# into a word; the length of a string is the number of symbols that it contains.\n",
    "# An example of a length 21 DNA string (whose alphabet contains the symbols 'A', 'C', 'G', and 'T') \n",
    "# is \"ATGCTTCAGAAAGGTCTTACG.\"\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "def count_nucleotides(dna):\n",
    "\n",
    "    dna_split = [*dna]\n",
    "\n",
    "    # check whether input dna is less than or equal to 1000 nucleotides\n",
    "    if(len(dna_split)<=1000):\n",
    "\n",
    "        values, counts = np.unique(dna_split, return_counts=True)\n",
    "        print(counts)   \n",
    "\n",
    "    else:\n",
    "        print('Sequence longer than the 1000 nucleotide limit.')\n",
    "\n",
    "\n",
    "count_nucleotides(\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. Enumerating Gene Orders\n",
    "# A permutation of length n is an ordering of the positive integers {1,2,…,n}\n",
    "# For example, π=(5,3,2,1,4) is a permutation of length 5\n",
    "# Given: A positive integer n≤7\n",
    "# Return: The total number of permutations of length n, followed by a list of all such permutations (in any order).\n",
    "\n",
    "import itertools\n",
    "\n",
    "def permute_n(n):\n",
    "    n_ls = list(range(1,n+1))\n",
    "    perms = list(itertools.permutations(range(1,n+1)))\n",
    "    len_perms = len(perms)\n",
    "    print(len_perms)\n",
    "    print(perms)\n",
    "\n",
    "permute_n(3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4. Getting the reverse complement\n",
    "# An RNA string is a string formed from the alphabet containing 'A', 'C', 'G', and 'U'.\n",
    "# Given a DNA string t corresponding to a coding strand, its transcribed RNA string u is \n",
    "# formed by replacing all occurrences of 'T' in t with 'U' in u. \n",
    "\n",
    "def rev_complement(dna):\n",
    "\n",
    "    # define dict of nucleotides\n",
    "    complement_dict = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "\n",
    "    # split input string into list of nucleotides and get the reverse\n",
    "    dna_split = [*dna]\n",
    "\n",
    "    # check whether input dna is less than or equal to 1000 nucleotides\n",
    "    if(len(dna_split)<=1000):\n",
    "\n",
    "        # lookup nucleotides in dict\n",
    "        comp_list = [complement_dict[x] for x in reversed(dna_split)]\n",
    "\n",
    "        # join into string and return\n",
    "        rna_string = ''.join(comp_list)\n",
    "        print(rna_string)\n",
    "    else:\n",
    "        print('Sequence longer than the 1000 nucleotide limit.')\n",
    "\n",
    "rev_complement('AGTACCGTGGTCTATCCACTGTCAGCCATCACGAGTACTTAAACGGAATGATATCCAAAAAGGGGGATGAACGTAACCTACGGTTCCCCCAGCTTGTCACAAGCTGTTTTCGGGTACATGCGTTCCCGATTATCGAGACTACGATAACTTTCATTGCCGACCTTCCTTGTTACATGGCGCGTGACACATAGCGGCGCCGGTGTGGTATCCATGACCGTTTCCGTACCATAGACGCGTGGGAGCTCCACGACCCTCGGAGTCTTCAGTAGAACGTGACTGTAATGCGCGATAACCACGCTTTTTTACCTACCAGCTTAACGCGCATGGTTTTTACATCTGCCCCCTTCGGTGCATCTCTAACAGAAGGACCACCAAGCGCCTACGCTGGAGATGTTAAGGAATTCAGGTCGGCCGTTACCCGGGGTAGATGATCTGTGGATCTTGCCGTTCCCAGCGGTGGAAACACATTCGTCATAGCACTTAGTGATCAGAGGTCACCCGCACATTAACCTTGGGCCGTCAGCATTGTGCACCTATCAGAGTAACACGATCTCTGCCGTGCTATGTTGTGTGGCGGCTTATAATTTTGCCAACACCCAGATGAGCCCTCTGGGTTTATCAGTGTGTAAATGAATTATCAAATGACATCAGGAAACTGTAACTCGGCGAGGCCGGTAGTCACACGTTACCCGTTTGCGGCGTGAGACAGTTGGATTGTCTGCACTACCATTCCCCTCCCGAAGTGTCTCGGCAGTGACCACGCTATTGCGACGACCGAAGTTATACTTGTTGCATATTATTCAAATGGATGGACAATCCATATAAGTGAATCATCTGAACAGGGGTAAACCGTTCTCTTGCAGCAGGTTAAACCAAAGAGAACTATGCAGGGCTAACGTCTAACGTCGTACTCGGTAATAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5. Rabbits and Recurrence Relations\n",
    "# A key observation is that the number of offspring in any month is equal to the number of rabbits that were alive two months prior.\n",
    "# a given month will contain rabbits alive previous month + offspring\n",
    "# number of offspring in any month is equal to the number of rabbits alive 2 months prior\n",
    "# return total number of pairs present after n months if we begin with 1 pair and every pair produces a litter of k pairs instead of 1 pair\n",
    "\n",
    "def fibonacci_rabbits(months, offspring):\n",
    "\n",
    "    parent, child = 1,1\n",
    "    for i in range(months-1):\n",
    "        child, parent = parent, parent + (child*offspring)\n",
    "    \n",
    "    return(child)\n",
    "\n",
    "fibonacci_rabbits(34, 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6. Computing GC Content\n",
    "# GC content is the % of symbols in the string that are C or G\n",
    "# Given 10 DNA strings in fasta format (length 1kbp each)\n",
    "# Return the ID of the string with the highest GC content \n",
    "\n",
    "# function to parse fasta file\n",
    "def fasta2dict(f):\n",
    "    lines = open(f, 'r').read().split('\\n>')\n",
    "    d={}\n",
    "    for i in range(len(lines)):\n",
    "        k = lines[i].split('\\n', 1)[0].replace('>', '')\n",
    "        v = lines[i].split('\\n', 1)[1].replace('\\n', '')\n",
    "        d[k] = v\n",
    "    return(d)\n",
    "\n",
    "def get_max_gc(d):\n",
    "\n",
    "    dict_gc = dict.fromkeys(d.keys(), 0)\n",
    "    for k,v in d.items():\n",
    "        dna = [*v]\n",
    "        perc_gc = ((dna.count('C') + dna.count('G'))/len(dna))*100\n",
    "        dict_gc[k] = perc_gc\n",
    "    dict_gc\n",
    "    print(max(dict_gc, key=dict_gc.get))\n",
    "    print(max(dict_gc.values()))\n",
    "\n",
    "fa_dict = fasta2dict('/Users/ainefairbrother/Documents/python_practice/rosalind_gc.txt')\n",
    "\n",
    "get_max_gc(fa_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7. Counting Point Mutations\n",
    "# How many bases differ between two sequences? \n",
    "# Return the Hamming difference\n",
    "\n",
    "def calc_hamming(seq1, seq2):\n",
    "\n",
    "    seq1 = [*seq1]\n",
    "    seq2 = [*seq2]\n",
    "\n",
    "    if(len(seq1)==len(seq2)):\n",
    "        score = [seq1[i]!=seq2[i] for i in range(len(seq1))].count(True)\n",
    "        return(score)\n",
    "    else:\n",
    "        print(\"Sequence 1 and sequence 2 not the same length\")\n",
    "\n",
    "calc_hamming(\n",
    "\"GCCGCCGACCTTGTCGTCGAAGGGATGACGGCAAGTTGACATACTCTCATCGGACATAGGCCAGTAGGCCTACTCCTTGCTTACCACGGGCATTCCCACCATGTAGCAGCCTTGCGACGACGTTGCCTCCGGAAAGTTACATACAATTGCGGACAATGATCGGTCCGCCCAAAAAATGATGAAGAATCGCTTGTAAGCTGGTTTCTGCTCCGCACCAGTCCGATCACTGGAGACGTCGTAGTCATCCACGTACTTGCTGATATTCTCCGCTGCGAATTATCCGCATCGGATAAAAAGTTAATGCCGTAGATCCGTCGGTGGCTTATGAATTGATCAACGTCCTGCATGACCGGGCTAAAGGGCTCGCCCCTCCACCCATACCGTTCCCCCTCCTCTCATACATCGTGAGCCGTGTATGGTTAAACGGCTCCCAAATATGAATAATCTAACAAGAGCCTGTCATTGGGTACTCTCTTACCAGTAGAGCGTGCCGGAGGTGCGCACCACAATTCACATTTGGCCAATGAATTTATATGCTGCCGATCATTATAACTATTCTTGAGCTAGACCCACCTATGGGTACCATCCTGTGAGCACCAGACGTTCAGAAGCCTGCTTAGACGACCTGCCATGCTCCGGATTTCGGAGCTCGAAGGATGGTCGGGCAATAGGCCTTCAATCTTCGGTTGTAGCAGTCAGGAGAACCCATCTATATTAGATCCTAACAGCCGACCATCGGTGAACGAGTGCCATCTCTACTTGCTGGCACTCTGTGATCAGACGTACACTACCCTGGCACGTTCCACGGCATGCCGAGGAATCTCCGCCCGAAAAGTGATGCGATTGCGCGCATGGCGGAGGATCGTTTGACTGTGGACCCCGAATTCTAGACCTTGTCTGTGATTTTCCTATATCCGCCCTCACTACCGTGCGGCAAAGGCAGACACTCATAGACTTGGACAAATGGCCCTAGGTGGTACCTCTGAACTCCTTAGATT\",\n",
    "\"AAAACCGACCAAAACGTGGATTTAAAAAAAGAGAGTTGTTGGAGACTAATTGGACATAAACGGATAGAACTGCCTCCCGGATGCCGTTGGCTCCCGCGGCCACTTGAGTCCAGTCGATGAGGGGGCATTAGTAGAGTGGCTTTCGTGTCTGGATTTTCCTCTATGCAGCCGGACACTCCTGCAGAGTCTGTTGCAACACTATAACTGCTCCAAAACCATCCGACTTCCGGTACCATTCAGAGCCTCTACGACCGCGTTGGCCCGCTTTGGCGCGCCTGCACCACTTCTGCAAGGCAGTTGTTCTCGCCTATCAGTGACACGCTAGTGAAAGGGGCTTACTTCGACCTGATAGTGGCACACCTCGTTTATCACGATGACTGGCGTGTCGGTCCCTTGCATCCCTCGTCACACGACAAGGTGTAAAACTTTGGCGAACGGGTAGACCTAAGCCATTGTCTTACTTGGGTACGCTTCATGACCATTCCGCGATCCGGAGGTATGCTGGAAGTCGTTCGTTTCCCAAGTGGAGTTAATTGGGTTATGTTTAGCCAGCCGACCGCGTATCATCTGACGATGGTCGAGCAATCGTATTAGCGCCCCGCGTTGAGAAACCATACTGGACGCCCTTCAAGGCCCAGGATATCGTACCGTTAAGGCGGGGAGTCCAGTAGGCTCTCCTACTCCGGTTTCGCGCGTCAATCACGATAAAGCATGTTCGTCTGAAACGACCGTCGTAAGGTAAAAAAGTTGTCTTTACATTGGGTCTCAGTAAGTCATCATAACCACACTAACCGGGCACTGTCTGGGGCGTCCGCCGGCAAATCGGACAGGGCCATGACGGAGTTCCTCGGATGCCGCCCGGTGCTCTCACTATAGAGTCCGATGCTGAAATCTGGCTACTGTCTGGCCGTCCTGGACATGCAATATCCTGCTACCAACGGATATGCTTTATAACGTTTTCAGATGGTTACAGGACTTGGCAACGTCAATCCGAAGTT\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 8. Finding a Motif in DNA\n",
    "# Given two DNA strings s and t, return all start locations of t as a substring of s\n",
    "\n",
    "def find_motif(s, t):\n",
    "\n",
    "    s_ls = [*s]\n",
    "    t_len = len(t)\n",
    "    start_indices = []\n",
    "\n",
    "    for i in range(len(s_ls)):\n",
    "        test = s_ls[i:i+t_len]\n",
    "        if ''.join(test) == t:\n",
    "            start_indices.append(i+1)\n",
    "\n",
    "    return(start_indices)\n",
    "\n",
    "starts = find_motif(\n",
    "    s=\"GAAGAATAATTGCCTCACTTGCCTCTTTTGCCTCAAATTGCCTCTTTTGCCTCATTGCCTCTTTGCCTCAACCTCAGTTGCCTCCTTGCCTCCCTTGCCTCGCGTTGCCTCGTAGGTTGCCTCATTGCCTCCAATTGCCTCTTTGCCTCTTGCCTCTCTATTTGCCTCTTGCCTCTTGCCTCTTGCCTCATTGCCTCATCACTTGCCTCTTGCCTCCCGATGTTGCCTCTTACGGCTTGCCTCGATGCTAAACTTGCCTCCTATTGCCTCTTGCCTCTTTGCCTCTTGCCTCATGTTGCCTCGATATTGCCTCTTGCCTCTTAATTGCCTCTTTGCCTCGGGTTGCCTCCCTCTTGCCTCCGTTGCCTCCACGGTTTGCCTCTTGCCTCTACGGTTGCCTCATGAATTGCCTCTGTTGCCTCACTATTGCCTCTACGGTTGCCTCTGCAGAGGCCTTGCCTCAATTGTTGCCTCGTTGCCTCGGAATGTTGCCTCGCAACGTTGCCTCATTGCCTCCTTTGCCTCTCTGATTGGAGTTTGCCTCCTTTAGTTGCCTCATTGCCTCTTGCCTCCGTCTTGCCTCCTTGCCTCTTTGCCTCATTTGCCTCCGCATTTGCCTCTATATTGCCTCTTGCCTCCGTTTGCCTCGCTTGCCTCACGTTTGCTCCTTGCCTCATTGCCTCTTGCCTCCTAGTTGCCTCCTTGCCTCGCTTGCCTCTCTTGCCTCTTGCCTCACCGCTTGCCTCTGGTTGCCTCCTTGCCTCTTGCCTCTTGCCTCATTTGCCTCGTTGCCTCGATTGCCTCATTACTTGCCTCTTTGCCTCTTGCCTCACAATTGCCTCAAGTCGATTTGCCTCTATATCCACGTTGCCTCATTGCCTCGTTTTATTGCCTCTGTTTGCCTCTATTGCCTCTTGCCTCCTTGCCTCTAATTGCCTCTTGCCTCCTTGCCTCTTGCCTCCTTGCCTCACATTTGCCTCTT\",\n",
    "    t=\"TTGCCTCTT\"\n",
    ")\n",
    "\n",
    "\" \".join(map(str, starts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 9. Mendel's First Law\n",
    "# Given 3 integers, k, m and n, representing a population of k+m+n organisms\n",
    "# k individuals - homozygous for dominant for a trait\n",
    "# m - heterozygous\n",
    "# n - homozygous recessive\n",
    "# The probability that two randomly selected mating organisms will produce a dominant allele \n",
    "\n",
    "from scipy.special import comb\n",
    "\n",
    "k = 18 \n",
    "m = 24 \n",
    "n = 19\n",
    "\n",
    "total_pop = k+m+n\n",
    "total_combs = comb(total_pop, 2)\n",
    "valid_combs = comb(k, 2) + k*m + k*n + 0.5*m*n + 0.75*comb(m, 2)\n",
    "probability = valid_combs/total_combs\n",
    "probability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 10. Finding a Shared Motif\n",
    "# Find the longest common substring of a set of strings\n",
    "\n",
    "from itertools import combinations\n",
    "\n",
    "# function to parse fasta file\n",
    "def fasta2dict(f):\n",
    "    lines = open(f, 'r').read().split('\\n>')\n",
    "    d={}\n",
    "    for i in range(len(lines)):\n",
    "        k = lines[i].split('\\n', 1)[0].replace('>', '')\n",
    "        v = lines[i].split('\\n', 1)[1].replace('\\n', '')\n",
    "        d[k] = v\n",
    "    return(d)\n",
    "\n",
    "def find_shared_motif(d):\n",
    "    collect_substrings = []\n",
    "    for i in range(len(d)):\n",
    "        entry = list(d.values())[i]\n",
    "        all_substrings = [entry[x:y] for x, y in combinations(range(len(entry) + 1), r = 2)]\n",
    "        collect_substrings.append(set(all_substrings))\n",
    "\n",
    "    return(max(set.intersection(*map(set, collect_substrings)), key=len))\n",
    "\n",
    "d = fasta2dict(\"/Users/ainefairbrother/Documents/python_practice/ex10.fa\")\n",
    "find_shared_motif(d)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 11. Translating RNA into Protein\n",
    "\n",
    "def rna_to_protein(rna):\n",
    "\n",
    "    # define codon dict\n",
    "    bases = \"UCAG\"\n",
    "    codons = [a + b + c for a in bases for b in bases for c in bases]\n",
    "    amino_acids = 'FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG'\n",
    "    codon_table = dict(zip(codons, amino_acids))\n",
    "\n",
    "    rna_codons = [rna[i:i+3] for i in range(0, len(rna), 3)]\n",
    "\n",
    "    protein = ''.join([codon_table[x] for x in rna_codons])\n",
    "    protein = protein.replace('*', '')\n",
    "\n",
    "    return(protein)\n",
    "\n",
    "rna_to_protein(\"AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 12. Consensus and Profile\n",
    "\n",
    "import numpy as np\n",
    "import sys\n",
    "\n",
    "# function to parse fasta file\n",
    "def fasta2dict(f):\n",
    "    lines = open(f, 'r').read().split('\\n>')\n",
    "    d={}\n",
    "    for i in range(len(lines)):\n",
    "        k = lines[i].split('\\n', 1)[0].replace('>', '')\n",
    "        v = lines[i].split('\\n', 1)[1].replace('\\n', '')\n",
    "        d[k] = v\n",
    "    return(d)\n",
    "\n",
    "# function to generate consensus string and profile matrix for sequences in a fasta file\n",
    "def consensus_and_profile(fa_path):\n",
    "\n",
    "    seqs = fasta2dict(fa_path)\n",
    "\n",
    "    seq_ls = list(seqs.values())\n",
    "\n",
    "    # get length of sequences, l\n",
    "    l = len(seq_ls[0])\n",
    "\n",
    "    collect_pos = []\n",
    "    for i in range(0,l):\n",
    "        pos_i = []\n",
    "        for seq in seq_ls:\n",
    "            pos_i.append(seq[i])\n",
    "        collect_pos.append(''.join(pos_i))\n",
    "        \n",
    "    collect_counts = []\n",
    "    for i in range(0, len(collect_pos)):\n",
    "        nuc_counts = [\n",
    "            collect_pos[i].count('A'), \n",
    "            collect_pos[i].count('C'), \n",
    "            collect_pos[i].count('G'), \n",
    "            collect_pos[i].count('T'), \n",
    "        ]\n",
    "        collect_counts.append(nuc_counts)\n",
    "\n",
    "    profile_matrix = np.array(collect_counts).transpose()\n",
    "\n",
    "    consensus_ls=[]\n",
    "    for i in range(0, len(collect_counts)):\n",
    "        # 0 is A\n",
    "        # 1 is C\n",
    "        # 2 is G\n",
    "        # 3 is T\n",
    "        max_index = collect_counts[i].index(max(collect_counts[i]))\n",
    "        if(max_index==0):\n",
    "            consensus_ls.append('A')\n",
    "        if(max_index==1):\n",
    "            consensus_ls.append('C')\n",
    "        if(max_index==2):\n",
    "            consensus_ls.append('G')\n",
    "        if(max_index==3):\n",
    "            consensus_ls.append('T')\n",
    "\n",
    "    consensus_string = ''.join(consensus_ls)\n",
    "\n",
    "    # check consensus = len of input\n",
    "    # print(len(consensus_string)==l)\n",
    "\n",
    "    print(\n",
    "    consensus_string,\"\\n\",\n",
    "    f\"A: {' '.join(map(str, profile_matrix[0]))}\\n \"\n",
    "    f\"C: {' '.join(map(str, profile_matrix[1]))}\\n \"\n",
    "    f\"G: {' '.join(map(str, profile_matrix[2]))}\\n \"\n",
    "    f\"T: {' '.join(map(str, profile_matrix[3]))} \"\n",
    "    )\n",
    "\n",
    "consensus_and_profile(\"/Users/ainefairbrother/Documents/python_practice/rosalind/ex12_sample.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 13. Open Reading Frames\n",
    "\n",
    "# 6 total reading framed\n",
    "# ORF - starts with start codon, ends with stop codon AND without any stop codons in between\n",
    "# candidate protein string is derived by translating an ORF until a stop codon is reached\n",
    "\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "def fasta2dict(f):\n",
    "    lines = open(f, 'r').read().split('\\n>')\n",
    "    d={}\n",
    "    for i in range(len(lines)):\n",
    "        k = lines[i].split('\\n', 1)[0].replace('>', '')\n",
    "        v = lines[i].split('\\n', 1)[1].replace('\\n', '')\n",
    "        d[k] = v\n",
    "    return(d)\n",
    "\n",
    "def find_orfs(input):\n",
    "\n",
    "    # define DNA and RNA complementary dicts\n",
    "    dna_comp_dict = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "    rna_dna_dict = {'A':'A', 'T':'U', 'C':'C', 'G':'G'}\n",
    "\n",
    "    # define RNA>codon dict\n",
    "    codon_dict = {'UUU':'F', 'UUC':'F', 'UUU':'F', 'UUC':'F', 'UUA':'L', 'UUG':'L', 'UCU':'S', 'UCC':'S', 'UCA':'S', 'UCG':'S', \n",
    "                'UAU':'Y', 'UAC':'Y', 'UAA':'Stop', 'UAG':'Stop', 'UGU':'C', 'UGC':'C', 'UGA':'Stop', 'UGG':'W', 'CUU':'L', \n",
    "                'CUC':'L', 'CUA':'L', 'CUG':'L', 'CCU':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P', 'CAU':'H','CAC':'H', 'CAA':'Q',\n",
    "                'CAG':'Q','CGU':'R','CGC':'R','CGA':'R','CGG':'R', 'AUU':'I', 'AUC':'I','AUA':'I','AUG':'M','ACU':'T','ACC':'T',\n",
    "                'ACA':'T','ACG':'T','AAU':'N','AAC':'N','AAA':'K','AAG':'K','AGU':'S','AGC':'S','AGA':'R','AGG':'R','GUU':'V',\n",
    "                'GUC':'V', 'GUA':'V','GUG':'V','GCU':'A','GCC':'A','GCA':'A','GCG':'A','GAU':'D','GAC':'D','GAA':'E','GAG':'E',\n",
    "                'GGU':'G','GGC':'G','GGA':'G','GGG':'G'}\n",
    "\n",
    "    # define start/stop codons\n",
    "    start = ['ATG']\n",
    "    stop = ['TAG', 'TGA', 'TAA']\n",
    "\n",
    "    def get_proteins(dna, frame, direction):\n",
    "            \n",
    "        if direction==\"fwd\":\n",
    "            dna = dna\n",
    "        elif direction==\"rev\":\n",
    "            dna = ''.join(reversed([dna_comp_dict[x] for x in [*dna]]))\n",
    "            \n",
    "        # truncate in line with frame 0, 1 or 2\n",
    "        dna = dna[frame:len(dna)]\n",
    "\n",
    "        # use pattern matching to find orfs\n",
    "        pattern = re.compile(r'(?=(ATG(?:...)*?)(?=TAG|TGA|TAA))')\n",
    "        finds = re.findall(pattern, dna)\n",
    "\n",
    "        # transcribe\n",
    "        finds_transcribed = [''.join([rna_dna_dict[x] for x in [*find]]) for find in finds]\n",
    "\n",
    "        # translate\n",
    "        proteins = []\n",
    "        for find_transcribed in finds_transcribed:\n",
    "            codons = [find_transcribed[i:i+3] for i in range(0, len(find_transcribed), 3)]\n",
    "            protein = ''.join([codon_dict[codon] for codon in codons])\n",
    "            proteins.append(protein)\n",
    "        \n",
    "        return(proteins)\n",
    "\n",
    "    # get proteins for fwd, rev and all reading frames 0,1,2\n",
    "    all_proteins = [get_proteins(input, 0, 'fwd'),\n",
    "    get_proteins(input, 1, 'fwd'),\n",
    "    get_proteins(input, 2, 'fwd'),\n",
    "    get_proteins(input, 0, 'rev'),\n",
    "    get_proteins(input, 1, 'rev'),\n",
    "    get_proteins(input, 2, 'rev')]\n",
    "    \n",
    "    # collapse list of lists\n",
    "    all_proteins = [item for sublist in all_proteins for item in sublist]\n",
    "\n",
    "    # print all unique proteins\n",
    "    for protein in set(all_proteins):\n",
    "        print(protein)\n",
    "\n",
    "find_orfs(\"TATGATGCTTAATATCGTGCCAAAATGTCGGACTTCTTCAGAAGTGACGGATATGATCCCTCGTGGTATAGTGTAGCAAGATCATAATGTAATGGTATTTTGCTGGCTCAAGGCTTGTTGGCGGTTCCCGATTAAGTATTTATCCTCTTGGATTGGTCGTATCTTCGTATCACGGTAAACGACTAAGAGCGGGCGGTGGCGTATGGACAACGTAAGCGAGAAAGAGGATGGAAGTGACTAACCCGCTAATACAAGTATGTAGCCTATTAGGCCCCCGCGCGTTAATCGAAATACATTTATAGATAGGTAGCGGAGACTTAACTGCATCCGTACGCCGCAGACTAAGAGGTGGTGATCTGTATACTGTAAAAAAAACAGTGATGAGAAGGTCAATACTAATCTCAAAAATCTACCCTCCGATCAGAGATTACCCTAAGGGGGCAATGCAAAGGCTGTGTCCGAATGCTCTAATATAGCTATATTAGAGCATTCGGACACAGCCTTTGCATTAATTAAACTAGTTGCACTGGCTATAGGAAGATTCGCGGGTCACCGGACGGAGAGGATGGTTGTGATCACACAACAAGTGTCCCTAGTGAATACCCTCGAGTAGCTTTATAGGCCCGGGAGTCCCACATAATACGGAACCCGGTCGGTGCTGGTCGACTAAAATCAGCCATGTCGTCCCGCCGAGACCATCGTCAGAAGCCGTTCACAAAGAGGGGTACCGTTCCTCGTCGACGGGTCTTACCGCTCTATAGTCCTAATAAGTTGCGCAGAAATCGTGTGAAGGGGAAACCCGGGGGTGCCAGTGTTATCCAATTAGGTCTGCATCGTGTAGACCAGCATCACGATAAAGGGTTTTGATAGGACTTGTACATCCGGACATCTAGCTGCTCCTAAAACTGACGGCACATAACTGCTCTATAATCCCGCTTCTGCATTTTTATAA\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
